mixins/anchor-manager.js

```diff
@@ -0,0 +1,169 @@
+// AnchorManagerMixin is a React mixin that can be used to make the
+// component support permalinkable anchors (URL fragments) that
+// attract the user's attention to them when navigated to.
+//
+// Any component that uses this mixin supports an `anchorId` prop,
+// which is the URL fragment (without the leading hash) of the anchor.
+// This is expected to be constant throughout the lifetime of the
+// component. (If `anchorId` is undefined, the mixin will be largely
+// disabled.)
+//
+// When the window's location hash matches the component's anchor id,
+// either at component mount time or during the `hashchange` window
+// event, the component's `attractAttentionToAnchor` state will be
+// set to `true` for 4 seconds. Additionally, if the component
+// defines a `handleAttractAttentionToAnchor` method, it will be
+// called. (The `attractAttentionToAnchor` method can be used to
+// manually trigger this behavior as well.)
+//
+// The 4-second default can be changed if the component defines
+// an instance variable called `ATTRACT_ATTENTION_TO_ANCHOR_DURATION`,
+// which should be set to an integer value in milliseconds.
+//
+// Note that this mixin will *not* directly affect the component's
+// rendering in any way. This means that it is not responsible for e.g.:
+//
+// * Setting any CSS classes, transitions, or animations.
+// * Setting the `id` attribute of any elements.
+// * Creating permalinks.
+
+var React = require('react');
+
+var DEFAULT_ATTRACT_DURATION = 4000;
+
+// AnchorManager is a largely private singleton class that maintains a
+// registry mapping anchor names to React components and listens for
+// location hash changes.
+//
+// When a registered component needs to be navigated to, its
+// `handleNavigateToAnchor` method is called; when it needs to
+// be navigated from, its `handleNavigateFromAnchor` method is called.
+var AnchorManager = function() {
+  this.anchorMap = {};
+  this.initialized = false;
+  this.currentAnchor = null;
+  this.handleHashChange = this.handleHashChange.bind(this);
+};
+
+AnchorManager.prototype = {
+  register: function(component) {
+    var anchorId = component.props.anchorId;
+    if (process.env.NODE_ENV !== 'production' &&
```

Comment:
Oh, because it's just a "sanity check" intended for development, and we don't want it to negatively impact production code. React's own [prop validation](http://facebook.github.io/react/docs/reusable-components.html#prop-validation) has similar behavior:

> Note that for performance reasons `propTypes` is only checked in development mode.